packages/jsii/lib/compiler.ts

```diff
@@ -495,23 +497,43 @@ export class Compiler implements Emitter {
    * This makes it so that running 'tsc' and running 'jsii' has the same behavior.
    */
   private determineSources(files: string[]): string[] {
-    const ret = new Array<string>();
+    const ret = new Set<string>();
 
     if (files.length > 0) {
-      ret.push(...files);
+      for (const file of files) {
+        ret.add(file);
+      }
     } else {
       const parseConfigHost = parseConfigHostFromCompilerHost(
         this.compilerHost,
       );
-      const parsed = ts.parseJsonConfigFileContent(
+      const { fileNames } = ts.parseJsonConfigFileContent(
         this.typescriptConfig,
         parseConfigHost,
         this.options.projectInfo.projectRoot,
       );
-      ret.push(...parsed.fileNames);
+      for (const file of fileNames) {
+        ret.add(file);
+      }
     }
 
-    return ret;
+    // Bonus: ensure all dependencies' entry points are included in the compiler
+    // input path. This guarantees we have symbols for all types, from the
+    // module root, which is necessary in order to properly detect submodules.
+    for (const assm of this.options.projectInfo.dependencyClosure) {
+      const { resolvedModule } = ts.resolveModuleName(
+        assm.name,
+        path.join(this.options.projectInfo.projectRoot, this.options.projectInfo.types),
+        this.typescriptConfig?.compilerOptions ?? {},
+        ts.sys,
+      );
+      if (!resolvedModule) {
+        continue;
+      }
+      ret.add(resolvedModule.resolvedFileName);
```

Comment:
Are the filenames from the config also resolved? If not, the deduplication probably won't work properly.

Reply:
Great question :) I had to check this, too... They are indeed resolved.